### PR TITLE
[Vue] Vue frontend doesn't detect custom primary key names

### LINKED ISSUE #23848: Vue frontend doesn't detect custom primary key names
##### **Overview of the issue**

In case we use a primary key with a name other then id the vuejs frontend incorrect generated.

##### **Motivation for or Use Case**
n.a.

##### **Reproduce the error**

Generate a new project using the JDL provided in "JHipster configuration" section, start the project, go to “Employees” entity and you wont be able to see the id.

##### **Related issues**
n.a.

##### **Suggest a Fix**
Change the files responsible for generating the vuejs frontened, to mach the appropriated key. 

It should target the
- src/main/webapp/app/entities/entity/*
- src/main/webapp/app/shared/model/entity.model.ts
- src/test/javascript/spec/app/entities/entity/*

##### **JHipster Version(s)**
[Release 7.9.4](https://www.jhipster.tech/2023/09/05/jhipster-release-7.9.4.html) (2023-09-05)

##### **JHipster configuration**
The following JDL can be use to demonstrate the issue

  <details>
  <summary>JDL definitions</summary>
  <pre>
application {
  config {
    baseName primarykeynonid,
    reactive true
    applicationType monolith
    authenticationType jwt
    buildTool maven
    packageName com.jhipster.demo.primarykeynonid,
    clientFramework vue
    prodDatabaseType mssql,
    enableTranslation true
    nativeLanguage pt-pt
    languages [en, pt-pt]
  }
  entities *
}

@dto(mapstruct)
@service(serviceClass)
entity Employee(FUNC1) {
    /**
     * The employee number
     */
    @fieldNameAsDatabaseColumn(NFUNC)
    @Id number String required unique
    /**
     * The employee name
     */
    @fieldNameAsDatabaseColumn(NOME)
    name String required
}

  </pre>
  </details>

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**
n.a.

##### **Browsers and Operating System**
n.a.

### PR DESCRIPTION
**Pull request by qmonmert**
Fix #23848 

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by mshima** [comment]
Integration test can be added by adding a vue-additional like:
https://github.com/jhipster/generator-jhipster/blob/872c49c4f0c5a80afb27aa64b97142c5651992ec/test-integration/workflow-samples/angular.json#L9-L16

Using `"jdl-entity": "custom-id.jdl"`.

**Comment by qmonmert** [comment]
ok thanks, more work than I expected

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_.vue.ejs [hunk 1]
```diff
@@ -173,7 +173,7 @@
                 <span id="<%= i18nKeyPrefix %>.delete.question" data-cy="<%= entityInstance %>DeleteDialogHeading" v-text="t$('entity.delete.title')">Confirm delete operation</span>
             </template>
             <div class="modal-body">
-                <p id="<%= jhiPrefixDashed %>-delete-<%= entityInstance %>-heading" v-text="t$('<%= i18nKeyPrefix %>.delete.question', {'id': removeId})">Are you sure you want to delete this <%= entityClassHumanized %>?</p>
+                <p id="<%= jhiPrefixDashed %>-delete-<%= entityInstance %>-heading" v-text="t$('<%= i18nKeyPrefix %>.delete.question', {'<%= primaryKey.name %>': removeId})">Are you sure you want to delete this <%= entityClassHumanized %>?</p>
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_.vue.ejs:176]: i18n variable.
```suggestion
                <p id="<%= jhiPrefixDashed %>-delete-<%= entityInstance %>-heading" v-text="t$('<%= i18nKeyPrefix %>.delete.question', {'id': removeId})">Are you sure you want to delete this <%= entityClassHumanized %>?</p>
```

**Comment by mshima** [comment]
I can add integration test if you want.

**Comment by qmonmert** [comment]
yes you can

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-details.vue.ejs [hunk 2]
```diff
@@ -49,20 +49,20 @@
                     <dd>
     <%_ if (relationship.otherEntityUser) { _%>
       <%_ if (relationship.relationshipManyToMany) { _%>
-                        <span v-for="(<%= relationshipFieldName %>, i) in <%= entityInstance %>.<%= relationshipFieldNamePlural %>" :key="<%= relationshipFieldName %>.id">{{i > 0 ? ', ' : ''}}
+                        <span v-for="(<%= relationshipFieldName %>, i) in <%= entityInstance %>.<%= relationshipFieldNamePlural %>" :key="<%= relationshipFieldName %>.<%= primaryKey.name %>">{{i > 0 ? ', ' : ''}}
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-details.vue.ejs:52]:
```suggestion
                        <span v-for="(<%= relationshipFieldName %>, i) in <%= entityInstance %>.<%= relationshipFieldNamePlural %>" :key="<%= relationshipFieldName %>.<%= relationship.otherEntity.primaryKey.name %>">{{i > 0 ? ', ' : ''}}
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-details.vue.ejs [hunk 3]
```diff
@@ -49,20 +49,20 @@
...
       <%_ } else { _%>
                         {{<%= entityInstance + "." + relationshipFieldName %> ? <%= entityInstance + "." + relationshipFieldName + "." + otherEntityField%> : ''}}
       <%_ } _%>
     <%_ } else { _%>
       <%_ if (relationship.relationshipManyToMany) { _%>
-                        <span v-for="(<%= relationshipFieldName %>, i) in <%= entityInstance %>.<%= relationshipFieldNamePlural %>" :key="<%= relationshipFieldName %>.id">{{i > 0 ? ', ' : ''}}
-                            <router-link :to="{name: '<%= otherEntityAngularName %>View', params: {<%= otherEntityName %>Id: <%= relationshipFieldName %>.id}}">{{<%= relationshipFieldName + "." + otherEntityField %>}}</router-link>
+                        <span v-for="(<%= relationshipFieldName %>, i) in <%= entityInstance %>.<%= relationshipFieldNamePlural %>" :key="<%= relationshipFieldName %>.<%= primaryKey.name %>">{{i > 0 ? ', ' : ''}}
+                            <router-link :to="{name: '<%= otherEntityAngularName %>View', params: {<%= otherEntityName %>Id: <%= relationshipFieldName %>.<%= primaryKey.name %>}}">{{<%= relationshipFieldName + "." + otherEntityField %>}}</router-link>
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-details.vue.ejs:61]:
```suggestion
                        <span v-for="(<%= relationshipFieldName %>, i) in <%= entityInstance %>.<%= relationshipFieldNamePlural %>" :key="<%= relationshipFieldName %>.<%= relationship.otherEntity.primaryKey.name %>">{{i > 0 ? ', ' : ''}}
                            <router-link :to="{name: '<%= otherEntityAngularName %>View', params: {<%= otherEntityName %>Id: <%= relationshipFieldName %>.<%= relationship.otherEntity.primaryKey.name %>}}">{{<%= relationshipFieldName + "." + otherEntityField %>}}</router-link>
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-details.vue.ejs [hunk 4]
```diff
@@ -49,20 +49,20 @@
...
                         </span>
       <%_ } else { _%>
                         <div v-if="<%= entityInstance + '.' + relationshipFieldName %>">
-                            <router-link :to="{name: '<%= otherEntityAngularName %>View', params: {<%= otherEntityName %>Id: <%= entityInstance + "." + relationshipFieldName %>.id}}">{{<%= entityInstance + "." + relationshipFieldName + "." + otherEntityField %>}}</router-link>
+                            <router-link :to="{name: '<%= otherEntityAngularName %>View', params: {<%= otherEntityName %>Id: <%= entityInstance + "." + relationshipFieldName %>.<%= primaryKey.name %>}}">{{<%= entityInstance + "." + relationshipFieldName + "." + otherEntityField %>}}</router-link>
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-details.vue.ejs:65]:
```suggestion
                            <router-link :to="{name: '<%= otherEntityAngularName %>View', params: {<%= otherEntityName %>Id: <%= entityInstance + "." + relationshipFieldName %>.<%= relationship.otherEntity.primaryKey.name %>}}">{{<%= entityInstance + "." + relationshipFieldName + "." + otherEntityField %>}}</router-link>
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-details.component.spec.ts.ejs [hunk 5]
```diff
@@ -83,7 +83,7 @@ describe('Component Tests', () => {
         <%= entityInstance %>ServiceStub.find.resolves(<%= entityInstance %>Sample);
         route = {
           params: {
-            <%= entityInstance %>Id: '' + <%- tsKeyId %>,
+            <%= entityInstance %><%= primaryKey.nameCapitalized %>: '' + <%- tsKeyId %>,
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-details.component.spec.ts.ejs:86]:
```suggestion
            <%= entityInstance %>Id: '' + <%- tsKeyId %>,
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.component.spec.ts.ejs [hunk 6]
```diff
@@ -178,7 +178,7 @@ describe('Component Tests', () => {
         // WHEN
         route = {
           params: {
-            <%= entityInstance %>Id: '' + <%= entityInstance %>Sample.<%- primaryKey.name %>,
+            <%= entityInstance %><%= primaryKey.nameCapitalized %>: '' + <%= entityInstance %>Sample.<%- primaryKey.name %>,
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.component.spec.ts.ejs:181]:
```suggestion
            <%= entityInstance %>Id: '' + <%= entityInstance %>Sample.<%- primaryKey.name %>,
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.component.ts.ejs [hunk 7]
```diff
@@ -242,7 +242,7 @@ _%>
   <%_ if (relationshipsContainManyToOne) { _%>
     getSelected(selectedVals, option) : any {
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.component.ts.ejs:243]: Probably need to be split

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.vue.ejs [hunk 8]
```diff
@@ -139,14 +139,14 @@ _%>
     <%_ } else { _%>
                             <option v-if="!<%= entityInstance %>.<%= relationshipName %>" v-bind:value="null" selected></option>
     <%_ } _%>
-                            <option v-bind:value="<%= entityInstance %>.<%=relationshipFieldName %> && <%=otherEntityName %>Option.id === <%= entityInstance %>.<%=relationshipFieldName %>.id ? <%= entityInstance %>.<%=relationshipFieldName %> : <%=otherEntityName %>Option" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.id">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
+                            <option v-bind:value="<%= entityInstance %>.<%=relationshipFieldName %> && <%=otherEntityName %>Option.<%= primaryKey.name %> === <%= entityInstance %>.<%=relationshipFieldName %>.<%= primaryKey.name %> ? <%= entityInstance %>.<%=relationshipFieldName %> : <%=otherEntityName %>Option" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.<%= primaryKey.name %>">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.vue.ejs:142]:
```suggestion
                            <option v-bind:value="<%= entityInstance %>.<%=relationshipFieldName %> && <%=otherEntityName %>Option.<%= relationship.otherEntity.primaryKey.name %> === <%= entityInstance %>.<%=relationshipFieldName %>.<%= relationship.otherEntity.primaryKey.name %> ? <%= entityInstance %>.<%=relationshipFieldName %> : <%=otherEntityName %>Option" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.<%= relationship.otherEntity.primaryKey.name %>">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.vue.ejs [hunk 9]
```diff
@@ -139,14 +139,14 @@ _%>
...
                         <select class="form-control" id="<%= entityFileName %>-<%= relationshipFieldNamePlural %>" data-cy="<%= relationshipFieldName %>" multiple name="<%= relationshipName %>" v-if="<%=entityInstance %>.<%=relationshipFieldNamePlural %> !== undefined" v-model="<%=entityInstance %>.<%=relationshipFieldNamePlural %>"<% if (relationshipRequired) { %> required<% } %>>
-                            <option v-bind:value="getSelected(<%=entityInstance %>.<%=relationshipFieldNamePlural %>, <%=otherEntityName %>Option)" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.id">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
+                            <option v-bind:value="getSelected(<%=entityInstance %>.<%=relationshipFieldNamePlural %>, <%=otherEntityName %>Option)" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.<%= primaryKey.name %>">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.vue.ejs:149]:
```suggestion
                            <option v-bind:value="getSelected(<%=entityInstance %>.<%=relationshipFieldNamePlural %>, <%=otherEntityName %>Option)" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.<%= relationship.otherEntity.primaryKey.name %>">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.component.ts.ejs (hunk 7 above)
Comment (mshima) [reply generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.component.ts.ejs:243]:
```suggestion
    getSelected(selectedVals, option, pkField = 'id') : any {
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.component.ts.ejs [hunk 10]
```diff
@@ -242,7 +242,7 @@ _%>
...
       if (selectedVals) {
-        return selectedVals.find(value => option.id === value.id) ?? option;
+        return selectedVals.find(value => option.<%= primaryKey.name %> === value.<%= primaryKey.name %>) ?? option;
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.component.ts.ejs:245]:
```suggestion
        return selectedVals.find(value => option[pkField] === value[pkField]) ?? option;
```

File: generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.vue.ejs [hunk 11]
```diff
@@ -139,14 +139,14 @@ _%>
...
                         <select class="form-control" id="<%= entityFileName %>-<%= relationshipFieldNamePlural %>" data-cy="<%= relationshipFieldName %>" multiple name="<%= relationshipName %>" v-if="<%=entityInstance %>.<%=relationshipFieldNamePlural %> !== undefined" v-model="<%=entityInstance %>.<%=relationshipFieldNamePlural %>"<% if (relationshipRequired) { %> required<% } %>>
-                            <option v-bind:value="getSelected(<%=entityInstance %>.<%=relationshipFieldNamePlural %>, <%=otherEntityName %>Option)" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.id">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
+                            <option v-bind:value="getSelected(<%=entityInstance %>.<%=relationshipFieldNamePlural %>, <%=otherEntityName %>Option)" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.<%= relationship.otherEntity.primaryKey.name %>">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
```
Comment (mshima) [inline generators/vue/templates/src/main/webapp/app/entities/_entityFolder_/_entityFile_-update.vue.ejs:149]:
```suggestion
                            <option v-bind:value="getSelected(<%=entityInstance %>.<%=relationshipFieldNamePlural %>, <%=otherEntityName %>Option, '<%= relationship.otherEntity.primaryKey.name %>')" v-for="<%=otherEntityName %>Option in <%=otherEntityNamePlural %>" :key="<%=otherEntityName %>Option.<%= relationship.otherEntity.primaryKey.name %>">{{<%=otherEntityName %>Option.<%=otherEntityField %>}}</option>
```

**Review by mshima — APPROVED** [review]

**Pull request merged**